[Bug] mysql-cdc 2.5-SNAPSHOT, failed to deserialize MAP event for full_types table

### Search before asking

- [X] I searched in the [issues](https://github.com/ververica/flink-cdc-connectors/issues) and found nothing similar.


### Flink version

1.18

### Flink CDC version

2.5-SNAPSHOT

### Database and its version

mysql  Ver 8.0.33 for macos13.2 on arm64 (Homebrew)


### Minimal reproduce step

# Create Table

```sql
Create Table: CREATE TABLE `test_table_1` (
  `id` bigint unsigned NOT NULL DEFAULT '0',
  `tiny_c` tinyint DEFAULT NULL,
  `tiny_un_c` tinyint unsigned DEFAULT NULL,
  `tiny_un_z_c` tinyint(3) unsigned zerofill DEFAULT NULL,
  `small_c` smallint DEFAULT NULL,
  `small_un_c` smallint unsigned DEFAULT NULL,
  `small_un_z_c` smallint(5) unsigned zerofill DEFAULT NULL,
  `medium_c` mediumint DEFAULT NULL,
  `medium_un_c` mediumint unsigned DEFAULT NULL,
  `medium_un_z_c` mediumint(8) unsigned zerofill DEFAULT NULL,
  `int_c` int DEFAULT NULL,
  `int_un_c` int unsigned DEFAULT NULL,
  `int_un_z_c` int(10) unsigned zerofill DEFAULT NULL,
  `int11_c` int DEFAULT NULL,
  `big_c` bigint DEFAULT NULL,
  `big_un_c` bigint unsigned DEFAULT NULL,
  `big_un_z_c` bigint(20) unsigned zerofill DEFAULT NULL,
  `varchar_c` varchar(255) CHARACTER SET utf8mb3 DEFAULT NULL,
  `char_c` char(3) CHARACTER SET utf8mb3 DEFAULT NULL,
  `real_c` double DEFAULT NULL,
  `float_c` float DEFAULT NULL,
  `float_un_c` float unsigned DEFAULT NULL,
  `float_un_z_c` float unsigned zerofill DEFAULT NULL,
  `double_c` double DEFAULT NULL,
  `double_un_c` double unsigned DEFAULT NULL,
  `double_un_z_c` double unsigned zerofill DEFAULT NULL,
  `decimal_c` decimal(8,4) DEFAULT NULL,
  `decimal_un_c` decimal(8,4) unsigned DEFAULT NULL,
  `decimal_un_z_c` decimal(8,4) unsigned zerofill DEFAULT NULL,
  `numeric_c` decimal(6,0) DEFAULT NULL,
  `big_decimal_c` decimal(65,1) DEFAULT NULL,
  `bit1_c` bit(1) DEFAULT NULL,
  `tiny1_c` tinyint(1) DEFAULT NULL,
  `boolean_c` tinyint(1) DEFAULT NULL,
  `date_c` date DEFAULT NULL,
  `time_c` time DEFAULT NULL,
  `datetime3_c` datetime(3) DEFAULT NULL,
  `datetime6_c` datetime(6) DEFAULT NULL,
  `timestamp_c` timestamp NULL DEFAULT NULL,
  `file_uuid` binary(16) DEFAULT NULL,
  `bit_c` bit(64) DEFAULT NULL,
  `text_c` text CHARACTER SET utf8mb3,
  `tiny_blob_c` tinyblob,
  `blob_c` blob,
  `medium_blob_c` mediumblob,
  `long_blob_c` longblob,
  `year_c` year DEFAULT NULL,
  `enum_c` enum('red','white') CHARACTER SET utf8mb3 DEFAULT 'red',
  `set_c` set('a','b') CHARACTER SET utf8mb3 DEFAULT NULL,
  `json_c` json DEFAULT NULL,
  `point_c` point DEFAULT NULL,
  `geometry_c` geometry DEFAULT NULL,
  `linestring_c` linestring DEFAULT NULL,
  `polygon_c` polygon DEFAULT NULL,
  `multipoint_c` multipoint DEFAULT NULL,
  `multiline_c` multilinestring DEFAULT NULL,
  `multipolygon_c` multipolygon DEFAULT NULL,
  `geometrycollection_c` geomcollection DEFAULT NULL,
  PRIMARY KEY (`id`)
)
```


# Configure flink job
```java
StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironment();

env.enableCheckpointing(10_000L);
StreamTableEnvironment tenv = StreamTableEnvironment.create(env);

tenv.executeQuery("CREATE TABLE test_table_1 (`id` DECIMAL(20,0),`tiny_c` TINYINT,`tiny_un_c` SMALLINT,`tiny_un_z_c` SMALLINT,`small_c` SMALLINT,`small_un_c` INT,`small_un_z_c` INT,`medium_c` INT,`medium_un_c` BIGINT,`medium_un_z_c` BIGINT,`int_c` INT,`int_un_c` BIGINT,`int_un_z_c` BIGINT,`int11_c` INT,`big_c` BIGINT,`big_un_c` DECIMAL(20,0),`big_un_z_c` DECIMAL(20,0),`varchar_c` STRING,`char_c` STRING,`real_c` DOUBLE,`float_c` FLOAT,`float_un_c` FLOAT,`float_un_z_c` FLOAT,`double_c` DOUBLE,`double_un_c` DOUBLE,`double_un_z_c` DOUBLE,`decimal_c` DECIMAL(8,4),`decimal_un_c` DECIMAL(8,4),`decimal_un_z_c` DECIMAL(8,4),`numeric_c` DECIMAL(6,0),`big_decimal_c` DECIMAL(65,1),`bit1_c` BOOLEAN,`tiny1_c` BOOLEAN,`boolean_c` BOOLEAN,`date_c` DATE,`time_c` TIME,`datetime3_c` TIMESTAMP(3),`datetime6_c` TIMESTAMP(6),`timestamp_c` TIMESTAMP,`file_uuid` BYTES,`bit_c` BYTES,`text_c` STRING,`tiny_blob_c` BYTES,`blob_c` BYTES,`medium_blob_c` BYTES,`long_blob_c` BYTES,`year_c` STRING,`enum_c` STRING,`set_c` STRING,`json_c` STRING,`point_c` STRING,`geometry_c` STRING,`linestring_c` STRING,`polygon_c` STRING,`multipoint_c` STRING,`multiline_c` STRING,`multipolygon_c` STRING,`geometrycollection_c` STRING,PRIMARY KEY (id) NOT ENFORCED) WITH ( 'connector' = 'mysql-cdc','hostname' = '127.0.0.1','port' = '3306','username' = 'root','password' = '','database-name' = 'test','table-name' = 'test_table_1')");

tenv.executeQuery("SELECT * FROM test_table_1")
```


# Wait for Snapshotting Phase Finish

# Insert Incremental Data

```sql

INSERT INTO test_table_1
VALUES (DEFAULT, 27, 255, 255, 32767, 65535, 65535, 8388607, 16777215, 16777215, 2147483647,
        4294967295, 4294967295, 2147483647, 9223372036854775807,
        18446744073709551615, 18446744073709551615,
        'Hello World', 'abc', 123.102, 123.102, 123.103, 123.104, 404.4443, 404.4444, 404.4445,
        123.4567, 123.4568, 123.4569, 345.6, 34567892.1, 0, 1, true,
        '2020-07-17', '18:00:22', '2020-07-17 18:00:22.123', '2020-07-17 18:00:22.123456',
        '2020-07-17 18:00:22',
        unhex(replace('651aed08-390f-4893-b2f1-36923e7b7400', '-', '')),
        b'0000010000000100000001000000010000000100000001000000010000000100',
        'text', UNHEX(HEX(16)), UNHEX(HEX(16)), UNHEX(HEX(16)), UNHEX(HEX(16)), 2021,
        'red', 'a,b,a', '{
    "key1": "value1"
  }',
        ST_GeomFromText('POINT(1 1)'),
        ST_GeomFromText('POLYGON((1 1, 2 1, 2 2,  1 2, 1 1))'),
        ST_GeomFromText('LINESTRING(3 0, 3 3, 3 5)'),
        ST_GeomFromText('POLYGON((1 1, 2 1, 2 2,  1 2, 1 1))'),
        ST_GeomFromText('MULTIPOINT((1 1),(2 2))'),
        ST_GeomFromText('MultiLineString((1 1,2 2,3 3),(4 4,5 5))'),
        ST_GeomFromText('MULTIPOLYGON(((0 0, 10 0, 10 10, 0 10, 0 0)), ((5 5, 7 5, 7 7, 5 7, 5 5)))'),
        ST_GeomFromText('GEOMETRYCOLLECTION(POINT(10 10), POINT(30 30), LINESTRING(15 15, 20 20))'));


```


# GOT

### What did you expect to see?

incremental data record successfully sinked to target system

### What did you see instead?


java.lang.RuntimeException: SplitFetcher thread 0 received unexpected exception while polling the records
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.runOnce(SplitFetcher.java:168) ~[flink-connector-files-1.18.0.jar:1.18.0]
	at org.apache.flink.connector.base.source.reader.fetcher.SplitFetcher.run(SplitFetcher.java:117) ~[flink-connector-files-1.18.0.jar:1.18.0]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
Caused by: org.apache.kafka.connect.errors.ConnectException: An exception occurred in the change event producer. This connector will be stopped.
	at io.debezium.pipeline.ErrorHandler.setProducerThrowable(ErrorHandler.java:50) ~[debezium-core-1.9.7.Final.jar:1.9.7.Final]
	at com.ververica.cdc.connectors.mysql.debezium.task.context.MySqlErrorHandler.setProducerThrowable(MySqlErrorHandler.java:85) ~[flink-connector-mysql-cdc-2.5-SNAPSHOT.jar:2.5-SNAPSHOT]
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$ReaderThreadLifecycleListener.onEventDeserializationFailure(MySqlStreamingChangeEventSource.java:1553) ~[flink-connector-mysql-cdc-2.5-SNAPSHOT.jar:1.9.7.Final]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:1064) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:631) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:932) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	... 1 common frames omitted
Caused by: io.debezium.DebeziumException: Failed to deserialize data of EventHeaderV4{timestamp=1699588796000, eventType=TABLE_MAP, serverId=1, headerLength=19, dataLength=196, nextPosition=16345945, flags=0}
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource.wrap(MySqlStreamingChangeEventSource.java:1488) ~[flink-connector-mysql-cdc-2.5-SNAPSHOT.jar:1.9.7.Final]
	... 5 common frames omitted
Caused by: com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1699588796000, eventType=TABLE_MAP, serverId=1, headerLength=19, dataLength=196, nextPosition=16345945, flags=0}
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:341) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeTableMapEventData(EventDeserializer.java:313) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:237) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	at io.debezium.connector.mysql.MySqlStreamingChangeEventSource$1.nextEvent(MySqlStreamingChangeEventSource.java:259) ~[flink-connector-mysql-cdc-2.5-SNAPSHOT.jar:1.9.7.Final]
	at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:1051) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	... 3 common frames omitted
Caused by: java.io.IOException: Unsupported table metadata field type 0
	at com.github.shyiko.mysql.binlog.event.deserialization.TableMapEventMetadataDeserializer.deserialize(TableMapEventMetadataDeserializer.java:52) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	at com.github.shyiko.mysql.binlog.event.deserialization.TableMapEventDataDeserializer.deserialize(TableMapEventDataDeserializer.java:59) ~[flink-connector-mysql-cdc-2.5-SNAPSHOT.jar:0.27.2]
	at com.github.shyiko.mysql.binlog.event.deserialization.TableMapEventDataDeserializer.deserialize(TableMapEventDataDeserializer.java:37) ~[flink-connector-mysql-cdc-2.5-SNAPSHOT.jar:0.27.2]
	at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:335) ~[mysql-binlog-connector-java-0.27.2.jar:0.27.2]
	... 7 common frames omitted

### Anything else?

Caused by: com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1699588796000, eventType=TABLE_MAP, serverId=1, headerLength=19, dataLength=196, nextPosition=16345945, flags=0}

`Unsupported table metadata field type 0`.

This is the crux of the problem. 

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventDataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventDataDeserializer.java
@@ -77,6 +77,7 @@ public class TableMapEventDataDeserializer implements EventDataDeserializer<Tabl
                 case NEWDECIMAL:
                 case FLOAT:
                 case DOUBLE:
+                case YEAR:
                     count++;
                     break;
                 default:

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -1,0 +1,217 @@
+/*
+ * Copyright 2013 Stanley Shyiko
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.shyiko.mysql.binlog.event.deserialization;
+
+import com.github.shyiko.mysql.binlog.event.TableMapEventMetadata;
+import com.github.shyiko.mysql.binlog.event.TableMapEventMetadata.DefaultCharset;
+import com.github.shyiko.mysql.binlog.io.ByteArrayInputStream;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue. */
+public class TableMapEventMetadataDeserializer {
+
+    private final Logger logger = Logger.getLogger(getClass().getName());
+
+    public TableMapEventMetadata deserialize(
+            ByteArrayInputStream inputStream, int nColumns, int nNumericColumns)
+            throws IOException {
+        int remainingBytes = inputStream.available();
+        if (remainingBytes <= 0) {
+            return null;
+        }
+
+        TableMapEventMetadata result = new TableMapEventMetadata();
+
+        for (; remainingBytes > 0; inputStream.enterBlock(remainingBytes)) {
+            int code = inputStream.readInteger(1);
+
+            MetadataFieldType fieldType = MetadataFieldType.byCode(code);
+            if (fieldType == null) {
+                throw new IOException("Unsupported table metadata field type " + code);
+            }
+
+            int fieldLength = inputStream.readPackedInteger();
+
+            remainingBytes = inputStream.available();
+            inputStream.enterBlock(fieldLength);
+
+            switch (fieldType) {
+                case SIGNEDNESS:
+                    result.setSignedness(readBooleanList(inputStream, nNumericColumns));
+                    inputStream.skipToTheEndOfTheBlock();
+                    break;
+                case DEFAULT_CHARSET:
+                    result.setDefaultCharset(readDefaultCharset(inputStream));
+                    break;
+                case COLUMN_CHARSET:
+                    result.setColumnCharsets(readIntegers(inputStream));
+                    break;
+                case COLUMN_NAME:
+                    result.setColumnNames(readColumnNames(inputStream));
+                    break;
+                case SET_STR_VALUE:
+                    result.setSetStrValues(readTypeValues(inputStream));
+                    break;
+                case ENUM_STR_VALUE:
+                    result.setEnumStrValues(readTypeValues(inputStream));
+                    break;
+                case GEOMETRY_TYPE:
+                    result.setGeometryTypes(readIntegers(inputStream));
+                    break;
+                case SIMPLE_PRIMARY_KEY:
+                    result.setSimplePrimaryKeys(readIntegers(inputStream));
+                    break;
+                case PRIMARY_KEY_WITH_PREFIX:
+                    result.setPrimaryKeysWithPrefix(readIntegerPairs(inputStream));
+                    break;
+                case ENUM_AND_SET_DEFAULT_CHARSET:
+                    result.setEnumAndSetDefaultCharset(readDefaultCharset(inputStream));
+                    break;
+                case ENUM_AND_SET_COLUMN_CHARSET:
+                    result.setEnumAndSetColumnCharsets(readIntegers(inputStream));
+                case VISIBILITY:
+                    result.setVisibility(readBooleanList(inputStream, nColumns));
+                    break;
+                case UNKNOWN_METADATA_FIELD_TYPE:
+                    if (logger.isLoggable(Level.FINE)) {
+                        logger.fine("Received metadata field of unknown type");
+                    }
+                    break;
+                default:
+                    inputStream.enterBlock(remainingBytes);
+                    throw new IOException("Unsupported table metadata field type " + code);
+            }
+            remainingBytes -= fieldLength;
+        }
+        return result;
+    }
+
+    private static BitSet readBooleanList(ByteArrayInputStream inputStream, int length)
+            throws IOException {
+        BitSet result = new BitSet();
+        // according to MySQL internals the amount of storage required for N columns is INT((N+7)/8)
+        // bytes
+        byte[] bytes = inputStream.read((length + 7) >> 3);
+        for (int i = 0; i < length; ++i) {
+            if ((bytes[i >> 3] & (1 << (7 - (i % 8)))) != 0) {
+                result.set(i);
+            }
+        }
+        return result;
+    }
+
+    private static DefaultCharset readDefaultCharset(ByteArrayInputStream inputStream)
+            throws IOException {
+        TableMapEventMetadata.DefaultCharset result = new TableMapEventMetadata.DefaultCharset();
+        result.setDefaultCharsetCollation(inputStream.readPackedInteger());
+        Map<Integer, Integer> charsetCollations = readIntegerPairs(inputStream);
+        if (!charsetCollations.isEmpty()) {
+            result.setCharsetCollations(charsetCollations);
+        }
+        return result;
+    }
+
+    private static List<Integer> readIntegers(ByteArrayInputStream inputStream) throws IOException {
+        List<Integer> result = new ArrayList<Integer>();
+        while (inputStream.available() > 0) {
+            result.add(inputStream.readPackedInteger());
+        }
+        return result;
+    }
+
+    private static List<String> readColumnNames(ByteArrayInputStream inputStream)
+            throws IOException {
+        List<String> columnNames = new ArrayList<String>();
+        while (inputStream.available() > 0) {
+            columnNames.add(inputStream.readLengthEncodedString());
+        }
+        return columnNames;
+    }
+
+    private static List<String[]> readTypeValues(ByteArrayInputStream inputStream)
+            throws IOException {
+        List<String[]> result = new ArrayList<String[]>();
+        while (inputStream.available() > 0) {
+            List<String> typeValues = new ArrayList<String>();
+            int valuesCount = inputStream.readPackedInteger();
+            for (int i = 0; i < valuesCount; ++i) {
+                typeValues.add(inputStream.readLengthEncodedString());
+            }
+            result.add(typeValues.toArray(new String[typeValues.size()]));
+        }
+        return result;
+    }
+
+    private static Map<Integer, Integer> readIntegerPairs(ByteArrayInputStream inputStream)
+            throws IOException {
+        Map<Integer, Integer> result = new LinkedHashMap<Integer, Integer>();
+        while (inputStream.available() > 0) {
+            int columnIndex = inputStream.readPackedInteger();
+            int columnCharset = inputStream.readPackedInteger();
+            result.put(columnIndex, columnCharset);
+        }
+        return result;
+    }
+
+    private enum MetadataFieldType {
+        SIGNEDNESS(1), // Signedness of numeric colums
+        DEFAULT_CHARSET(2), // Charsets of character columns
+        COLUMN_CHARSET(3), // Charsets of character columns
+        COLUMN_NAME(4), // Names of columns
+        SET_STR_VALUE(5), // The string values of SET columns
+        ENUM_STR_VALUE(6), // The string values is ENUM columns
+        GEOMETRY_TYPE(7), // The real type of geometry columns
+        SIMPLE_PRIMARY_KEY(8), // The primary key without any prefix
+        PRIMARY_KEY_WITH_PREFIX(9), // The primary key with some prefix
+        ENUM_AND_SET_DEFAULT_CHARSET(10), // Charsets of ENUM and SET columns
+        ENUM_AND_SET_COLUMN_CHARSET(11), // Charsets of ENUM and SET columns
+        VISIBILITY(12), // Column visibility (8.0.23 and newer)
+        UNKNOWN_METADATA_FIELD_TYPE(
+                128); // Returned with binlog-row-metadata=FULL from MySQL 8.0 in some cases
+
+        private final int code;
+
+        private MetadataFieldType(int code) {
+            this.code = code;
+        }
+
+        public int getCode() {
+            return code;
+        }
+
+        private static final Map<Integer, MetadataFieldType> INDEX_BY_CODE;
+
+        static {
+            INDEX_BY_CODE = new HashMap<Integer, MetadataFieldType>();
+            for (MetadataFieldType fieldType : values()) {
+                INDEX_BY_CODE.put(fieldType.code, fieldType);
+            }
+        }
+
+        public static MetadataFieldType byCode(int code) {
+            return INDEX_BY_CODE.get(code);
+        }
+    }
+}

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -13,6 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
 package com.github.shyiko.mysql.binlog.event.deserialization;
 
 import com.github.shyiko.mysql.binlog.event.TableMapEventMetadata;
@@ -91,6 +92,8 @@ public class TableMapEventMetadataDeserializer {
                     break;
                 case ENUM_AND_SET_COLUMN_CHARSET:
                     result.setEnumAndSetColumnCharsets(readIntegers(inputStream));
+                    result.setVisibility(readBooleanList(inputStream, nColumns));
+                    break;
                 case VISIBILITY:
                     result.setVisibility(readBooleanList(inputStream, nColumns));
                     break;

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -30,7 +30,9 @@ import java.util.Map;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
-/** Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue. */
+/** Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue #363.
+ * Issue link: https://github.com/shyiko/mysql-binlog-connector-java/issues/363
+ * */
 public class TableMapEventMetadataDeserializer {
 
     private final Logger logger = Logger.getLogger(getClass().getName());

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -31,7 +31,7 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 
 /**
- * Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue #363.
+ * Copied from mysql-binlog-connector 0.27.2 to fix table metadata deserialization issue #363.
  *
  * <p>Line 74: skip to end of the signedness block so that it won't affect parsing the next table metadata.
  */

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -31,16 +31,18 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 
 /**
- * Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue #363. Line
- * 66: skip to end of the signedness block so that it won't affect parsing the next table metadata.
+ * Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue #363.
+ *
+ * <p>Line 74: skip to end of the signedness block so that it won't affect parsing the next table metadata.
+ */
+/**
+ * @author <a href="mailto:ahmedahamid@yahoo.com">Ahmed Abdul Hamid</a>
  */
 public class TableMapEventMetadataDeserializer {
 
     private final Logger logger = Logger.getLogger(getClass().getName());
 
-    public TableMapEventMetadata deserialize(
-            ByteArrayInputStream inputStream, int nColumns, int nNumericColumns)
-            throws IOException {
+    public TableMapEventMetadata deserialize(ByteArrayInputStream inputStream, int nColumns, int nNumericColumns) throws IOException {
         int remainingBytes = inputStream.available();
         if (remainingBytes <= 0) {
             return null;
@@ -54,6 +56,12 @@ public class TableMapEventMetadataDeserializer {
             MetadataFieldType fieldType = MetadataFieldType.byCode(code);
             if (fieldType == null) {
                 throw new IOException("Unsupported table metadata field type " + code);
+            }
+            if (MetadataFieldType.UNKNOWN_METADATA_FIELD_TYPE.equals(fieldType)) {
+                if (logger.isLoggable(Level.FINE)) {
+                    logger.fine("Received metadata field of unknown type");
+                }
+                continue;
             }
 
             int fieldLength = inputStream.readPackedInteger();
@@ -95,15 +103,8 @@ public class TableMapEventMetadataDeserializer {
                     break;
                 case ENUM_AND_SET_COLUMN_CHARSET:
                     result.setEnumAndSetColumnCharsets(readIntegers(inputStream));
-                    result.setVisibility(readBooleanList(inputStream, nColumns));
-                    break;
                 case VISIBILITY:
                     result.setVisibility(readBooleanList(inputStream, nColumns));
-                    break;
-                case UNKNOWN_METADATA_FIELD_TYPE:
-                    if (logger.isLoggable(Level.FINE)) {
-                        logger.fine("Received metadata field of unknown type");
-                    }
                     break;
                 default:
                     inputStream.enterBlock(remainingBytes);
@@ -114,11 +115,9 @@ public class TableMapEventMetadataDeserializer {
         return result;
     }
 
-    private static BitSet readBooleanList(ByteArrayInputStream inputStream, int length)
-            throws IOException {
+    private static BitSet readBooleanList(ByteArrayInputStream inputStream, int length) throws IOException {
         BitSet result = new BitSet();
-        // according to MySQL internals the amount of storage required for N columns is INT((N+7)/8)
-        // bytes
+        // according to MySQL internals the amount of storage required for N columns is INT((N+7)/8) bytes
         byte[] bytes = inputStream.read((length + 7) >> 3);
         for (int i = 0; i < length; ++i) {
             if ((bytes[i >> 3] & (1 << (7 - (i % 8)))) != 0) {
@@ -128,8 +127,7 @@ public class TableMapEventMetadataDeserializer {
         return result;
     }
 
-    private static DefaultCharset readDefaultCharset(ByteArrayInputStream inputStream)
-            throws IOException {
+    private static DefaultCharset readDefaultCharset(ByteArrayInputStream inputStream) throws IOException {
         TableMapEventMetadata.DefaultCharset result = new TableMapEventMetadata.DefaultCharset();
         result.setDefaultCharsetCollation(inputStream.readPackedInteger());
         Map<Integer, Integer> charsetCollations = readIntegerPairs(inputStream);
@@ -147,8 +145,7 @@ public class TableMapEventMetadataDeserializer {
         return result;
     }
 
-    private static List<String> readColumnNames(ByteArrayInputStream inputStream)
-            throws IOException {
+    private static List<String> readColumnNames(ByteArrayInputStream inputStream) throws IOException {
         List<String> columnNames = new ArrayList<String>();
         while (inputStream.available() > 0) {
             columnNames.add(inputStream.readLengthEncodedString());
@@ -156,8 +153,7 @@ public class TableMapEventMetadataDeserializer {
         return columnNames;
     }
 
-    private static List<String[]> readTypeValues(ByteArrayInputStream inputStream)
-            throws IOException {
+    private static List<String[]> readTypeValues(ByteArrayInputStream inputStream) throws IOException {
         List<String[]> result = new ArrayList<String[]>();
         while (inputStream.available() > 0) {
             List<String> typeValues = new ArrayList<String>();
@@ -170,8 +166,7 @@ public class TableMapEventMetadataDeserializer {
         return result;
     }
 
-    private static Map<Integer, Integer> readIntegerPairs(ByteArrayInputStream inputStream)
-            throws IOException {
+    private static Map<Integer, Integer> readIntegerPairs(ByteArrayInputStream inputStream) throws IOException {
         Map<Integer, Integer> result = new LinkedHashMap<Integer, Integer>();
         while (inputStream.available() > 0) {
             int columnIndex = inputStream.readPackedInteger();
@@ -182,20 +177,19 @@ public class TableMapEventMetadataDeserializer {
     }
 
     private enum MetadataFieldType {
-        SIGNEDNESS(1), // Signedness of numeric colums
-        DEFAULT_CHARSET(2), // Charsets of character columns
-        COLUMN_CHARSET(3), // Charsets of character columns
-        COLUMN_NAME(4), // Names of columns
-        SET_STR_VALUE(5), // The string values of SET columns
-        ENUM_STR_VALUE(6), // The string values is ENUM columns
-        GEOMETRY_TYPE(7), // The real type of geometry columns
-        SIMPLE_PRIMARY_KEY(8), // The primary key without any prefix
-        PRIMARY_KEY_WITH_PREFIX(9), // The primary key with some prefix
-        ENUM_AND_SET_DEFAULT_CHARSET(10), // Charsets of ENUM and SET columns
-        ENUM_AND_SET_COLUMN_CHARSET(11), // Charsets of ENUM and SET columns
-        VISIBILITY(12), // Column visibility (8.0.23 and newer)
-        UNKNOWN_METADATA_FIELD_TYPE(
-                128); // Returned with binlog-row-metadata=FULL from MySQL 8.0 in some cases
+        SIGNEDNESS(1),                      // Signedness of numeric colums
+        DEFAULT_CHARSET(2),                 // Charsets of character columns
+        COLUMN_CHARSET(3),                  // Charsets of character columns
+        COLUMN_NAME(4),                     // Names of columns
+        SET_STR_VALUE(5),                   // The string values of SET columns
+        ENUM_STR_VALUE(6),                  // The string values is ENUM columns
+        GEOMETRY_TYPE(7),                   // The real type of geometry columns
+        SIMPLE_PRIMARY_KEY(8),              // The primary key without any prefix
+        PRIMARY_KEY_WITH_PREFIX(9),         // The primary key with some prefix
+        ENUM_AND_SET_DEFAULT_CHARSET(10),   // Charsets of ENUM and SET columns
+        ENUM_AND_SET_COLUMN_CHARSET(11),    // Charsets of ENUM and SET columns
+        VISIBILITY(12),                     // Column visibility (8.0.23 and newer)
+        UNKNOWN_METADATA_FIELD_TYPE(128);   // Returned with binlog-row-metadata=FULL from MySQL 8.0 in some cases
 
         private final int code;
 
@@ -203,9 +197,7 @@ public class TableMapEventMetadataDeserializer {
             this.code = code;
         }
 
-        public int getCode() {
-            return code;
-        }
+        public int getCode() { return code; }
 
         private static final Map<Integer, MetadataFieldType> INDEX_BY_CODE;
 

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -1,11 +1,11 @@
 /*
- * Copyright 2013 Stanley Shyiko
+ * Copyright 2023 Ververica Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- * http://www.apache.org/licenses/LICENSE-2.0
+ *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -31,7 +31,7 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 
 /** Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue #363.
- * Issue link: https://github.com/shyiko/mysql-binlog-connector-java/issues/363
+ * Line 66: skip to end of the signedness block so that it won't affect parsing the next table metadata.
  * */
 public class TableMapEventMetadataDeserializer {
 

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -33,16 +33,16 @@ import java.util.logging.Logger;
 /**
  * Copied from mysql-binlog-connector 0.27.2 to fix table metadata deserialization issue #363.
  *
- * <p>Line 74: skip to end of the signedness block so that it won't affect parsing the next table metadata.
- */
-/**
- * @author <a href="mailto:ahmedahamid@yahoo.com">Ahmed Abdul Hamid</a>
+ * <p>Line 75: skip to end of the signedness block so that it won't affect parsing the next table
+ * metadata.
  */
 public class TableMapEventMetadataDeserializer {
 
     private final Logger logger = Logger.getLogger(getClass().getName());
 
-    public TableMapEventMetadata deserialize(ByteArrayInputStream inputStream, int nColumns, int nNumericColumns) throws IOException {
+    public TableMapEventMetadata deserialize(
+            ByteArrayInputStream inputStream, int nColumns, int nNumericColumns)
+            throws IOException {
         int remainingBytes = inputStream.available();
         if (remainingBytes <= 0) {
             return null;
@@ -103,6 +103,8 @@ public class TableMapEventMetadataDeserializer {
                     break;
                 case ENUM_AND_SET_COLUMN_CHARSET:
                     result.setEnumAndSetColumnCharsets(readIntegers(inputStream));
+                    result.setVisibility(readBooleanList(inputStream, nColumns));
+                    break;
                 case VISIBILITY:
                     result.setVisibility(readBooleanList(inputStream, nColumns));
                     break;
@@ -115,9 +117,11 @@ public class TableMapEventMetadataDeserializer {
         return result;
     }
 
-    private static BitSet readBooleanList(ByteArrayInputStream inputStream, int length) throws IOException {
+    private static BitSet readBooleanList(ByteArrayInputStream inputStream, int length)
+            throws IOException {
         BitSet result = new BitSet();
-        // according to MySQL internals the amount of storage required for N columns is INT((N+7)/8) bytes
+        // according to MySQL internals the amount of storage required for N columns is INT((N+7)/8)
+        // bytes
         byte[] bytes = inputStream.read((length + 7) >> 3);
         for (int i = 0; i < length; ++i) {
             if ((bytes[i >> 3] & (1 << (7 - (i % 8)))) != 0) {
@@ -127,7 +131,8 @@ public class TableMapEventMetadataDeserializer {
         return result;
     }
 
-    private static DefaultCharset readDefaultCharset(ByteArrayInputStream inputStream) throws IOException {
+    private static DefaultCharset readDefaultCharset(ByteArrayInputStream inputStream)
+            throws IOException {
         TableMapEventMetadata.DefaultCharset result = new TableMapEventMetadata.DefaultCharset();
         result.setDefaultCharsetCollation(inputStream.readPackedInteger());
         Map<Integer, Integer> charsetCollations = readIntegerPairs(inputStream);
@@ -145,7 +150,8 @@ public class TableMapEventMetadataDeserializer {
         return result;
     }
 
-    private static List<String> readColumnNames(ByteArrayInputStream inputStream) throws IOException {
+    private static List<String> readColumnNames(ByteArrayInputStream inputStream)
+            throws IOException {
         List<String> columnNames = new ArrayList<String>();
         while (inputStream.available() > 0) {
             columnNames.add(inputStream.readLengthEncodedString());
@@ -153,7 +159,8 @@ public class TableMapEventMetadataDeserializer {
         return columnNames;
     }
 
-    private static List<String[]> readTypeValues(ByteArrayInputStream inputStream) throws IOException {
+    private static List<String[]> readTypeValues(ByteArrayInputStream inputStream)
+            throws IOException {
         List<String[]> result = new ArrayList<String[]>();
         while (inputStream.available() > 0) {
             List<String> typeValues = new ArrayList<String>();
@@ -166,7 +173,8 @@ public class TableMapEventMetadataDeserializer {
         return result;
     }
 
-    private static Map<Integer, Integer> readIntegerPairs(ByteArrayInputStream inputStream) throws IOException {
+    private static Map<Integer, Integer> readIntegerPairs(ByteArrayInputStream inputStream)
+            throws IOException {
         Map<Integer, Integer> result = new LinkedHashMap<Integer, Integer>();
         while (inputStream.available() > 0) {
             int columnIndex = inputStream.readPackedInteger();
@@ -177,19 +185,20 @@ public class TableMapEventMetadataDeserializer {
     }
 
     private enum MetadataFieldType {
-        SIGNEDNESS(1),                      // Signedness of numeric colums
-        DEFAULT_CHARSET(2),                 // Charsets of character columns
-        COLUMN_CHARSET(3),                  // Charsets of character columns
-        COLUMN_NAME(4),                     // Names of columns
-        SET_STR_VALUE(5),                   // The string values of SET columns
-        ENUM_STR_VALUE(6),                  // The string values is ENUM columns
-        GEOMETRY_TYPE(7),                   // The real type of geometry columns
-        SIMPLE_PRIMARY_KEY(8),              // The primary key without any prefix
-        PRIMARY_KEY_WITH_PREFIX(9),         // The primary key with some prefix
-        ENUM_AND_SET_DEFAULT_CHARSET(10),   // Charsets of ENUM and SET columns
-        ENUM_AND_SET_COLUMN_CHARSET(11),    // Charsets of ENUM and SET columns
-        VISIBILITY(12),                     // Column visibility (8.0.23 and newer)
-        UNKNOWN_METADATA_FIELD_TYPE(128);   // Returned with binlog-row-metadata=FULL from MySQL 8.0 in some cases
+        SIGNEDNESS(1), // Signedness of numeric colums
+        DEFAULT_CHARSET(2), // Charsets of character columns
+        COLUMN_CHARSET(3), // Charsets of character columns
+        COLUMN_NAME(4), // Names of columns
+        SET_STR_VALUE(5), // The string values of SET columns
+        ENUM_STR_VALUE(6), // The string values is ENUM columns
+        GEOMETRY_TYPE(7), // The real type of geometry columns
+        SIMPLE_PRIMARY_KEY(8), // The primary key without any prefix
+        PRIMARY_KEY_WITH_PREFIX(9), // The primary key with some prefix
+        ENUM_AND_SET_DEFAULT_CHARSET(10), // Charsets of ENUM and SET columns
+        ENUM_AND_SET_COLUMN_CHARSET(11), // Charsets of ENUM and SET columns
+        VISIBILITY(12), // Column visibility (8.0.23 and newer)
+        UNKNOWN_METADATA_FIELD_TYPE(
+                128); // Returned with binlog-row-metadata=FULL from MySQL 8.0 in some cases
 
         private final int code;
 
@@ -197,7 +206,9 @@ public class TableMapEventMetadataDeserializer {
             this.code = code;
         }
 
-        public int getCode() { return code; }
+        public int getCode() {
+            return code;
+        }
 
         private static final Map<Integer, MetadataFieldType> INDEX_BY_CODE;
 

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventMetadataDeserializer.java
@@ -30,9 +30,10 @@ import java.util.Map;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
-/** Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue #363.
- * Line 66: skip to end of the signedness block so that it won't affect parsing the next table metadata.
- * */
+/**
+ * Copied from mysql-binlog-connector 0.25.3 to fix table metadata deserialization issue #363. Line
+ * 66: skip to end of the signedness block so that it won't affect parsing the next table metadata.
+ */
 public class TableMapEventMetadataDeserializer {
 
     private final Logger logger = Logger.getLogger(getClass().getName());

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventDataDeserializerTest.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/java/com/github/shyiko/mysql/binlog/event/deserialization/TableMapEventDataDeserializerTest.java
@@ -22,7 +22,9 @@ import com.github.shyiko.mysql.binlog.io.ByteArrayInputStream;
 import org.junit.Test;
 
 import java.io.IOException;
+import java.util.Arrays;
 import java.util.BitSet;
+import java.util.List;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
@@ -95,6 +97,66 @@ public class TableMapEventDataDeserializerTest {
         assertThat(eventData.toString()).isEqualTo(getExpectedEventData().toString());
     }
 
+    @Test
+    public void testDeserializeMetadata() throws IOException {
+        byte[] data = {
+            // SIGNEDNESS
+            1,
+            // SIGNEDNESS length
+            5,
+            // 5 bytes for SIGNEDNESS
+            -74,
+            -39,
+            -101,
+            97,
+            0,
+            // COLUMN_CHARSET
+            3,
+            // COLUMN_CHARSET length
+            32,
+            // remaining 32 bytes for COLUMN_CHARSET
+            33,
+            33,
+            63,
+            33,
+            63,
+            63,
+            63,
+            63,
+            -4,
+            -1,
+            0,
+            -4,
+            -1,
+            0,
+            -4,
+            -1,
+            0,
+            -4,
+            -1,
+            0,
+            -4,
+            -1,
+            0,
+            -4,
+            -1,
+            0,
+            -4,
+            -1,
+            0,
+            -4,
+            -1,
+            0
+        };
+
+        TableMapEventMetadataDeserializer deserializer = new TableMapEventMetadataDeserializer();
+
+        TableMapEventMetadata metadata =
+                deserializer.deserialize(new ByteArrayInputStream(data), 59, 32);
+
+        assertThat(metadata.toString()).isEqualTo(getExpectedTableMapEventMetaData().toString());
+    }
+
     private TableMapEventData getExpectedEventData() {
         TableMapEventData eventData = new TableMapEventData();
         // table_id
@@ -122,5 +184,24 @@ public class TableMapEventDataDeserializerTest {
         metadata.setDefaultCharset(charset);
         eventData.setEventMetadata(metadata);
         return eventData;
+    }
+
+    private TableMapEventMetadata getExpectedTableMapEventMetaData() {
+        // optional metadata fields
+        TableMapEventMetadata metadata = new TableMapEventMetadata();
+        BitSet signedness = new BitSet();
+
+        List<Integer> signedBits =
+                Arrays.asList(0, 2, 3, 5, 6, 8, 9, 11, 12, 15, 16, 19, 20, 22, 23, 25, 26, 31);
+
+        for (Integer signedBit : signedBits) {
+            signedness.set(signedBit);
+        }
+
+        metadata.setSignedness(signedness);
+        metadata.setColumnCharsets(
+                Arrays.asList(
+                        33, 33, 63, 33, 63, 63, 63, 63, 255, 255, 255, 255, 255, 255, 255, 255));
+        return metadata;
     }
 }

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/table/MySqlConnectorITCase.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/table/MySqlConnectorITCase.java
@@ -2219,8 +2219,8 @@ public class MySqlConnectorITCase extends MySqlSourceTestBase {
                                 + " 'server-id' = '%s',"
                                 + " 'scan.incremental.snapshot.chunk.size' = '%s'"
                                 + ")",
-                        MYSQL_CONTAINER.getHost(),
-                        MYSQL_CONTAINER.getDatabasePort(),
+                        MYSQL8_CONTAINER.getHost(),
+                        MYSQL8_CONTAINER.getDatabasePort(),
                         TEST_USER,
                         TEST_PASSWORD,
                         binlogDatabase.getDatabaseName(),
@@ -2242,16 +2242,15 @@ public class MySqlConnectorITCase extends MySqlSourceTestBase {
         try (Connection connection = binlogDatabase.getJdbcConnection();
                 Statement statement = connection.createStatement()) {
             statement.execute(
-                    "INSERT INTO binlog_metadata VALUES (1, 127, 255, 255, 32767, 65535, 65535, 2024),(2, 127, 255, 255, 32767, 65535, 65535, 2025);");
-            statement.execute("UPDATE binlog_metadata SET year_c=2025 WHERE id=1;");
-            statement.execute("DELETE FROM binlog_metadata WHERE id=2;");
+                    "INSERT INTO binlog_metadata VALUES (2, 127, 255, 255, 32767, 65535, 65535, 2024),(3, 127, 255, 255, 32767, 65535, 65535, 2024);");
+            statement.execute("DELETE FROM binlog_metadata WHERE id=3;");
         }
 
         String[] expected =
                 new String[] {
                     // snapshot records
-                    "+I[0, 127, 255, 255, 32767, 65535, 65535, 2023]",
-                    "+I[1, 127, 255, 255, 32767, 65535, 65535, 2025]"
+                    "+I[1, 127, 255, 255, 32767, 65535, 65535, 2023]",
+                    "+I[2, 127, 255, 255, 32767, 65535, 65535, 2024]"
                 };
         assertEqualsInAnyOrder(Arrays.asList(expected), fetchRows(iterator, expected.length));
         result.getJobClient().get().cancel().get();

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/table/MySqlConnectorITCase.java
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/table/MySqlConnectorITCase.java
@@ -106,6 +106,9 @@ public class MySqlConnectorITCase extends MySqlSourceTestBase {
     private final UniqueDatabase inventoryDatabase8 =
             new UniqueDatabase(MYSQL8_CONTAINER, "inventory", TEST_USER, TEST_PASSWORD);
 
+    private final UniqueDatabase binlogDatabase =
+            new UniqueDatabase(MYSQL8_CONTAINER, "binlog_metadata_test", TEST_USER, TEST_PASSWORD);
+
     private final StreamExecutionEnvironment env =
             StreamExecutionEnvironment.getExecutionEnvironment();
     private final StreamTableEnvironment tEnv =
@@ -2184,6 +2187,74 @@ public class MySqlConnectorITCase extends MySqlSourceTestBase {
                             + "2. The server id has been used by the mysql cdc table in other jobs.\n"
                             + "3. The server id has been used by other sync tools like canal, debezium and so on.\n");
         }
+    }
+
+    @Test
+    public void testBinlogTableMetadataDeserialization() throws Exception {
+        if (!incrementalSnapshot) {
+            return;
+        }
+        binlogDatabase.createAndInitialize();
+        String sourceDDL =
+                String.format(
+                        "CREATE TABLE binlog_metadata (\n"
+                                + "    id BIGINT NOT NULL,\n"
+                                + "    tiny_c TINYINT,\n"
+                                + "    tiny_un_c SMALLINT ,\n"
+                                + "    tiny_un_z_c SMALLINT ,\n"
+                                + "    small_c SMALLINT,\n"
+                                + "    small_un_c INT,\n"
+                                + "    small_un_z_c INT,\n"
+                                + "    year_c INT,\n"
+                                + " PRIMARY KEY(id) NOT ENFORCED"
+                                + ") WITH ("
+                                + " 'connector' = 'mysql-cdc',"
+                                + " 'hostname' = '%s',"
+                                + " 'port' = '%s',"
+                                + " 'username' = '%s',"
+                                + " 'password' = '%s',"
+                                + " 'database-name' = '%s',"
+                                + " 'table-name' = '%s',"
+                                + " 'server-time-zone' = 'UTC',"
+                                + " 'server-id' = '%s',"
+                                + " 'scan.incremental.snapshot.chunk.size' = '%s'"
+                                + ")",
+                        MYSQL_CONTAINER.getHost(),
+                        MYSQL_CONTAINER.getDatabasePort(),
+                        TEST_USER,
+                        TEST_PASSWORD,
+                        binlogDatabase.getDatabaseName(),
+                        "binlog_metadata",
+                        getServerId(),
+                        getSplitSize());
+        tEnv.executeSql(sourceDDL);
+
+        // async submit job
+        TableResult result = tEnv.executeSql("SELECT * FROM binlog_metadata");
+
+        // wait for the source startup, we don't have a better way to wait it, use sleep for now
+        do {
+            Thread.sleep(5000L);
+        } while (result.getJobClient().get().getJobStatus().get() != RUNNING);
+
+        CloseableIterator<Row> iterator = result.collect();
+
+        try (Connection connection = binlogDatabase.getJdbcConnection();
+                Statement statement = connection.createStatement()) {
+            statement.execute(
+                    "INSERT INTO binlog_metadata VALUES (1, 127, 255, 255, 32767, 65535, 65535, 2024),(2, 127, 255, 255, 32767, 65535, 65535, 2025);");
+            statement.execute("UPDATE binlog_metadata SET year_c=2025 WHERE id=1;");
+            statement.execute("DELETE FROM binlog_metadata WHERE id=2;");
+        }
+
+        String[] expected =
+                new String[] {
+                    // snapshot records
+                    "+I[0, 127, 255, 255, 32767, 65535, 65535, 2023]",
+                    "+I[1, 127, 255, 255, 32767, 65535, 65535, 2025]"
+                };
+        assertEqualsInAnyOrder(Arrays.asList(expected), fetchRows(iterator, expected.length));
+        result.getJobClient().get().cancel().get();
     }
 
     // ------------------------------------------------------------------------------------

--- a/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/resources/ddl/binlog_metadata_test.sql
+++ b/flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/test/resources/ddl/binlog_metadata_test.sql
@@ -1,0 +1,36 @@
+-- Copyright 2023 Ververica Inc.
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--   http://www.apache.org/licenses/LICENSE-2.0
+-- Unless required by applicable law or agreed to in writing,
+-- software distributed under the License is distributed on an
+-- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+-- KIND, either express or implied.  See the License for the
+-- specific language governing permissions and limitations
+-- under the License.
+
+-- ----------------------------------------------------------------------------------------------------------------
+-- DATABASE:  column_type_test
+-- ----------------------------------------------------------------------------------------------------------------
+
+-- create a table with that number of numeric columns x meets the condition 8|(x + 32), so the table contains
+-- 8 numeric columns. Which mean any miscalculation in numeric columns will result in the wrong number of bytes
+-- read when parsing the signedness table metadata.
+
+CREATE TABLE binlog_metadata
+(
+    id                   SERIAL,
+    tiny_c               TINYINT,
+    tiny_un_c            TINYINT UNSIGNED,
+    tiny_un_z_c          TINYINT UNSIGNED ZEROFILL,
+    small_c              SMALLINT,
+    small_un_c           SMALLINT UNSIGNED,
+    small_un_z_c         SMALLINT UNSIGNED ZEROFILL,
+    year_c YEAR,
+    PRIMARY KEY (id)
+) DEFAULT CHARSET=utf8;
+
+INSERT INTO binlog_metadata
+VALUES (DEFAULT, 127, 255, 255, 32767, 65535, 65535, 2023);

